packages/devtools_app/lib/src/ui/hover.dart

```diff
@@ -358,74 +368,139 @@ class _HoverCardTooltipState extends State<HoverCardTooltip> {
+          asyncTimeout: asyncTimeout,
         );
-
-        _setHoverCard(
-          spinnerHoverCard,
+      } else {
+        _setHoverCardFromData(
+          generateHoverCardData!(event),
+          context: context,
+          event: event,
         );
+      }
+    });
+  }
 
-        // The spinner is showing, we can now generate the HoverCardData
-        hoverCardData = await widget.asyncGenerateHoverCardData!(
+  void _showAsyncHoverCard({
+    required AsyncGenerateHoverCardDataFunc asyncGenerateHoverCardData,
+    required PointerHoverEvent event,
+    int? asyncTimeout,
+  }) async {
+    HoverCard? spinnerHoverCard;
+    final hoverCardDataFuture = asyncGenerateHoverCardData(
+      event: event,
+      isHoverStale: () =>
+          spinnerHoverCard != null &&
+          !_hoverCardController.isHoverCardStillActive(spinnerHoverCard),
+    );
+    // If we have set the async hover card to show up only after a timeout,
+    // then race the timeout against generating the hover card data. If
+    // generating the data completes first, immediately show the hover card
+    // (or return early if there is no data).
+    if (asyncTimeout != null) {
+      final hoverCardDataCompleter =
+          _hoverCardDataCompleter(hoverCardDataFuture);
+      await Future.any([
+        _timeoutCompleter(asyncTimeout).future,
+        hoverCardDataCompleter.future,
+      ]);
+
+      if (hoverCardDataCompleter.isCompleted) {
+        final data = await hoverCardDataCompleter.future;
+        // If we get no data back, then don't show a hover card.
+        if (data == null) return;
+        // Otherwise, show a hover card immediately.
+        return _setHoverCardFromData(
+          data,
+          context: context,
           event: event,
-          isHoverStale: () =>
-              !_hoverCardController.isHoverCardStillActive(spinnerHoverCard),
         );
+      }
+    }
+    // The data on the card is fetched asynchronously, so show a spinner
+    // while we wait for it.
+    spinnerHoverCard = HoverCard.fromHoverEvent(
+      context: context,
+      contents: const CenteredCircularProgressIndicator(),
+      width: HoverCardTooltip.defaultHoverWidth,
+      event: event,
+      hoverCardController: _hoverCardController,
+    );
 
-        if (!_hoverCardController.isHoverCardStillActive(spinnerHoverCard)) {
-          // The hovercard became stale while fetching it's data. So it should
-          // no longer be shown.
-          return;
-        }
-        if (hoverCardData == null) {
-          // No data was provided so remove the spinner
-          _removeHoverCard(spinnerHoverCard);
-          return;
-        }
-      } else {
-        assert(widget.generateHoverCardData != null);
+    _setHoverCard(
+      spinnerHoverCard,
+    );
 
-        hoverCardData = widget.generateHoverCardData!(event);
-      }
+    // The spinner is showing, we can now generate the HoverCardData
+    final hoverCardData = await hoverCardDataFuture;
```

Comment:
is it possible to wait on our completer instead? Something about this makes it seem like we would be fetching the data a 2nd time by calling this again

Reply:
Good point! Switched to using the completer